dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,248 @@
+4. Create a migration test as detailed in [migrations](apollo-router/src/configuration/migrations/README.md)
+5. In your migration description tell the users what they have to update.
+
+## Process
+It should be obvious to the user what they are configuring and how it will affect Router behaviour. It's tricky for us as developers to know when something isn't obvious to users as often we are too close to the domain.
+
+Complex config should be run by the rest of the team. Ideally before writing any code, as coming to the team late can cause code churn and frustration. The process is as follows:
+1. In the github issue put the proposed config in.
+2. List any concerns.
+3. Notify the team that you are looking for request for comment.
+4. Schedule a meeting to discuss. (This is important, often design considerations will fall out of conversation)
+5. If it is not completely clear what the direction should be:
+1. Wait a few days, often people will have ideas later even if they didn't in the meeting.
+2. Ask users what they think.
+6. Make your changes.
+
+Note that these are not hard and fast rules, and if your config is really obviously correct then bu all means make the change and be prepared to deal with comments at the review stage.
+
+## Design patterns
+
+Use the following as a rule of thumb, also look at existing config for inspiration.
+The most important goal is usability, so do break the rules if it makes sense, but it's worth bringing the discussion to the team in such circumstances.  
+
+1. [Avoid empty config](#avoid-empty-config).
+2. [Do use `#[serde(deny_unknown_fields)]`](#do-use-serdedeny_unknown_fields).
+3. [Don't use `#[serde(flatten)]`](#dont-use-serdeflatten).
+4. [Document your configuration options](#document-your-configuration-options).
+5. [Plan for the future](#plan-for-the-future).
+
+### Avoid empty config
+
+In Rust you can use `Option` to say that config is optional, however this can give a bad experience if the type is complex and all fields are optional.
+
+#### GOOD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url // url is required
+}
+```
+```yaml
+export:
+    url: http://example.com
+```
+
+#### GOOD
+```rust
+enum ExportUrl {
+    Default,
+    Url(Url)
+}
+
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: ExportUrl // Url is required but user may specify `default`
+}
+```
+```yaml
+export:
+    url: default
+```
+
+#### BAD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Optional<Url> // url is optional
+}
+```
+```yaml
+export: # The user is not aware that url was defaulted.
+```
+
+#### UGLY
+In the case where you genuinely have no config then it may be acceptable to have an `enabled: bool` flag.
+The reason to avoid this is that it creates a disconnect between the user ant the thing that they are trying to do
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    enabled: bool,
+    url: Optional<Url> // url is optional
+}
```

Comment:
I think sometimes it's not so ugly in case you have a default value for the optional value in yaml. You should mention that.
Let me explain why it's not ugly/bad when you have a default value.
+ Having to write always the default value when you just want to enable something could be boring.
+ Enforcing this default value could also be very useful in specific cases instead of asking to the users to write it by themselves just to enable something. The main reason is to have a kind of convention. Like if we know there is a default value, let's say a default header name. We can do some assumptions on that, for example if you're developing a graphql client which wants to be by default compliant with the router, it knows the default value and are able to be compliant for free, it's like a convention because it's the default value.

Reply:
The problem is the lack of consistency. There are lots of places in the config we could have an `enabled` field. Pretty much at every level

For instance telemetry export:
```yaml
telemetry:
  enabled: true # Enable or disable telemetry plugin completely
  tracing:
    enabled: true # Enable or disable tracing
    datadog: 
      enabled: true # Defaults to datadog agent at localhost. Endpoint can become optional now
    otlp: 
      enabled: true # Defaults to GRPC to localhost, Endpoint can become optional now
```

If we decide to go down the enabled route we should look at the rest of the config and put `enabled` in so that the user gets a consistent experience.

There is an advantage to using enabled across the board. If users want to enable or disable stuff via env variable they can, but I'm not completely convinced.

Reply:
This is the pattern we have used elsewhere:
https://github.com/apollographql/router/pull/2159/files/9093b125f43d3fe19ca3d5b28c055705b6f89a94#diff-2897bc49cf0aab22634663d6ae15fd678e61323c9a5a63e5a90003a9d0973bbdR60-R75

Reply:
Changed in
[2f3193a](https://github.com/apollographql/router/pull/2159/commits/2f3193a84cda2159342a1b4eceea68e0f6644f5e)